From 1f523daffaaef85f1252ee854688282a5a0a46be Mon Sep 17 00:00:00 2001
From: Bruno Martins <bgcngm@gmail.com>
Date: Wed, 5 Sep 2018 01:46:44 +0100
Subject: [PATCH 04/12] Settings: Check interfaces before enabling ADB over
 network

 * Make sure that an active connection is established in
   one of the supported interfaces - WiFi or USB network.

Change-Id: I6c5c468f1b58aefe64065c373968cae0e700e415
---
 res/values/cm_strings.xml                     |  1 +
 .../WirelessAdbPreferenceController.java      | 12 ++++++++-
 .../development/WirelessAdbWarningDialog.java | 25 +++++++++++++++----
 3 files changed, 32 insertions(+), 6 deletions(-)

diff --git a/res/values/cm_strings.xml b/res/values/cm_strings.xml
index e2dbb0d609..70f580240a 100644
--- a/res/values/cm_strings.xml
+++ b/res/values/cm_strings.xml
@@ -257,6 +257,7 @@
     <string name="adb_over_network">ADB over network</string>
     <string name="adb_over_network_summary">Enable TCP/IP debugging over network interfaces (Wi\u2011Fi, USB networks). This setting is reset on reboot</string>
     <string name="adb_over_network_warning">WARNING: When ADB over network is enabled, your phone is open for intrusions on all connected networks!\n\nOnly use this feature when you are connected on trusted networks.\n\nDo you really want to enable this function?</string>
+    <string name="adb_over_network_disallowed_error">ERROR: ADB over network can only be enabled when connected to a Wi\u2011Fi or USB network!</string>
 
     <!-- [CHAR LIMIT=NONE] Device Info screen. Countdown for user taps to enable development settings -->
     <plurals name="show_dev_countdown_cm">
diff --git a/src/com/android/settings/development/WirelessAdbPreferenceController.java b/src/com/android/settings/development/WirelessAdbPreferenceController.java
index e1e61e14d0..08b0befe30 100644
--- a/src/com/android/settings/development/WirelessAdbPreferenceController.java
+++ b/src/com/android/settings/development/WirelessAdbPreferenceController.java
@@ -17,6 +17,8 @@
 package com.android.settings.development;
 
 import android.content.Context;
+import android.net.ConnectivityManager;
+import android.net.NetworkInfo;
 import android.net.NetworkUtils;
 import android.net.wifi.WifiInfo;
 import android.net.wifi.WifiManager;
@@ -96,9 +98,17 @@ public class WirelessAdbPreferenceController extends DeveloperOptionsPreferenceC
         }
     }
 
+    private boolean isConnectedToSupportedInterface() {
+        ConnectivityManager connMgr = mContext.getSystemService(ConnectivityManager.class);
+        NetworkInfo wiredNwInfo = connMgr.getNetworkInfo(ConnectivityManager.TYPE_ETHERNET);
+        NetworkInfo wirelessNwInfo = connMgr.getNetworkInfo(ConnectivityManager.TYPE_WIFI);
+        return (wiredNwInfo != null && wiredNwInfo.isConnected()) ||
+                (wirelessNwInfo != null && wirelessNwInfo.isConnected());
+    }
+
     public void onWirelessAdbDialogConfirmed() {
         LineageSettings.Secure.putInt(mContext.getContentResolver(),
-                LineageSettings.Secure.ADB_PORT, 5555);
+                LineageSettings.Secure.ADB_PORT, !isConnectedToSupportedInterface() ? -1 : 5555);
         updatePreference();
     }
 
diff --git a/src/com/android/settings/development/WirelessAdbWarningDialog.java b/src/com/android/settings/development/WirelessAdbWarningDialog.java
index 2297baa9fc..e37f6934da 100644
--- a/src/com/android/settings/development/WirelessAdbWarningDialog.java
+++ b/src/com/android/settings/development/WirelessAdbWarningDialog.java
@@ -21,6 +21,8 @@ import android.app.Dialog;
 import android.app.Fragment;
 import android.app.FragmentManager;
 import android.content.DialogInterface;
+import android.net.ConnectivityManager;
+import android.net.NetworkInfo;
 import android.os.Bundle;
 
 import com.android.internal.logging.nano.MetricsProto;
@@ -49,11 +51,16 @@ public class WirelessAdbWarningDialog extends InstrumentedDialogFragment impleme
 
     @Override
     public Dialog onCreateDialog(Bundle savedInstanceState) {
-        return new AlertDialog.Builder(getActivity())
-                .setMessage(R.string.adb_over_network_warning)
-                .setPositiveButton(android.R.string.ok, this /* onClickListener */)
-                .setNegativeButton(android.R.string.cancel, this /* onClickListener */)
-                .create();
+        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
+        if (isConnectedToSupportedInterface()) {
+            builder.setMessage(R.string.adb_over_network_warning);
+            builder.setPositiveButton(android.R.string.ok, this /* onClickListener */);
+            builder.setNegativeButton(android.R.string.cancel, this /* onClickListener */);
+        } else {
+            builder.setMessage(R.string.adb_over_network_disallowed_error);
+            builder.setPositiveButton(android.R.string.ok, this /* onClickListener */);
+        }
+        return builder.create();
     }
 
     @Override
@@ -78,4 +85,12 @@ public class WirelessAdbWarningDialog extends InstrumentedDialogFragment impleme
         }
         host.onWirelessAdbDialogDismissed();
     }
+
+    private boolean isConnectedToSupportedInterface() {
+        ConnectivityManager connMgr = getActivity().getSystemService(ConnectivityManager.class);
+        NetworkInfo wiredNwInfo = connMgr.getNetworkInfo(ConnectivityManager.TYPE_ETHERNET);
+        NetworkInfo wirelessNwInfo = connMgr.getNetworkInfo(ConnectivityManager.TYPE_WIFI);
+        return (wiredNwInfo != null && wiredNwInfo.isConnected()) ||
+                (wirelessNwInfo != null && wirelessNwInfo.isConnected());
+    }
 }
-- 
2.17.1

